### DISCUSSION TITLE
How to tell build system to get the edited file from application directory instead of zephyr stack using CMakelists.txt

### DISCUSSION THREAD

**Discussion by aravinth-kalai** [zephyrproject-rtos/zephyr#67707] [Q&A]
Hello everyone !

I have edited the isotp.c file from zephyr stack , I want to keep the edited file in my local application directory, during west build the compiler should consider the isotp.c file from my application directory instead of zephyr directory, how can I edit my main application CMakeLists.txt file to make it happen

**Rough  application directory folder structure:**
```
main.c
zephyr/subsys/canbus/isotp/isotp.c -> I want this edited file needs to be considered during west build
CMakeLists.txt

```
**zephyr stack directory:**

```
subsys/canbus/isotp/isotp.c -> I don't want this file to be considered during west build 

```

Any suggestions or references link to understand this concept and make it happen using CMakeLists.txt 

Thanks in advance!

**Comment by nordicjm** [comment]
use `zephyr_library_amend()`, see https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/extensions.cmake#L431 for details

**Reply by aravinth-kalai** [reply]
Hi @nordicjm,

Thanks for your reply, I have tried the following two lines in my main CMakeLists.txt
```
zephyr_library_amend()
zephyr_library_sources(isotp.c)
```
and I'm keeping the file in a application directory be like 
```
zephyr/subsys/canbus/isotp/isotp.c
```
if I tried to build it I'm getting this error 
```
Function only available for Zephyr modules.
Call Stack (most recent call first):
  CMakeLists.txt:147 (zephyr_library_amend)

```

**Reply by nordicjm** [reply]
Yes it needs to be part of a zephyr module, make the part that changes the library file a module and it will work

**Reply by aravinth-kalai** [reply]
But this is not from zephyr/modules directory in this case what should I have to do

**Reply by nordicjm** [reply]
Has nothing to do with that directory, you need to make it a module as per https://docs.zephyrproject.org/latest/develop/modules.html

**Reply by borob-engineering** [reply]
Hello,
I tried the described way to change the behavior of the mcumanger handler of the FS group from inside my external project. The file I would like to change is zephyr\subsys\mgmt\mcumgr\grp\fs_mgmt\src\fs_mgmt.c
So I added the subsys folder structure including the mentioned file and the Makefile to my (external) project folder. The compilation works fine but during linking I get errors due to multiple definition of symbols.

The following two lines are causing the errors:
MCUMGR_HANDLER_DEFINE(fs_mgmt, fs_mgmt_register_group); **-> multiple definition of fs_mgmt**
LOG_MODULE_REGISTER(mcumgr_fs_grp, CONFIG_MCUMGR_GRP_FS_LOG_LEVEL); ****-> multiple definition of log_const_mcumgr_fs_grp**

Could you please give me a hint how to solve this?

Thanks
Robert

**Update: it seems that the during the zephyr specific compile run of fs_mgmt.c the symbols for logging ang mcu handler registration are generated indenpendently of the amend. So I commented out these two lines in my version of the code to avoid multiple symbol generation. After that the linking is working and my own code is executed as desired. Please let me know if there is a better way to solve this.**

**Reply by nordicjm** [reply]
you disable fs mgmt then add your own Kconfig to enable your replacement with a different Kconfig name

**Reply by borob-engineering** [reply]
> you disable fs mgmt then add your own Kconfig to enable your replacement with a different Kconfig name

Tested this, but this leads to compile errors due to unresolved header files in zephyr/mgmt/mcumgr/mgmt/callbacks.h which checks if CONFIG_MCUMGR_GRP_FS is set. As workaround I've done this in my implemetation of fs_mgmt.c:

    #define CONFIG_MCUMGR_GRP_FS
    #include <zephyr/mgmt/mcumgr/mgmt/callbacks.h>
    #undef CONFIG_MCUMGR_GRP_FS

This works and I see that my adapted code is executed. Any better way to avoid the unresolved header file?
